doc/specs/#2129 - Side-by-side.md

```diff
@@ -0,0 +1,100 @@
+---
+author: Demitrius Nelon @denelon
+created on: 2023-11-30
+last updated: 2023-11-30
+issue id: 2129
+---
+
+# Better support for Side by Side packages
+
+[comment]: # Link to issue: "For [#2929](https://github.com/microsoft/winget-cli/issues/2129)"
+
+## Abstract
+
+WinGet should better support multiple versions of the same package being installed on the same machine. This happens, and WinGet currently tries to pretend it doesn't.
+
+## Inspiration
+
+Multiple [issues](https://github.com/microsoft/winget-cli/issues?q=is%3Aissue+is%3Aopen+label%3Aside-by-side) have been submitted with the "side-by-side" label.
+
+## Solution Design
+
+Per comment by @JohnMcPMS:
+> Replace PackageVersion InstalledVersion with PackageVersion[] InstalledVersions internally, then reasoning about how that affects everywhere that breaks.
+
+## UI/UX Design
+
+Several different commands will be impacted by this change. In some cases, additional information will be presented to the user so they are aware of the implications for the command executed.
+
+### WinGet Install
+
+When one or more versions (but not the latest) of a package are installed on the system and none of them are pinned, running `winget install <package>` will install the latest version of the package side-by-side unless the manifest specifies the previous version should be uninstalled or denied. The user may override this behavior with "--retain-previous".
+> Note: WinGet will not attempt to override the default behavior the installer may have related to the removal of previous versions of the package.
+
+### WinGet List
+
+When more than one version of a package is installed on the system, all installed versions will all be listed. Only the latest version of an installed package will show results in the "Available" column if a newer version is available in a configured source.
+
+### WinGet Upgrade
+
+When more than one version of a package is installed on the system, only the latest installed version will be listed if a newer version is available in a configured source.
+
+### WinGet Uninstall
+
+For `winget uninstall <package>`:
+
+When more than one version of a package is installed on the system, the user will be prompted to disambiguate. The user will be informed "Multiple package versions found matching input criteria. Please specify the version to be uninstalled or use "--all" to remove all versions of the package". All installed versions will be displayed so the user may either disambiguate by specifying the version to be removed or by passing "--all" as an argument to specify every installed version of the package should be uninstalled.

```

Comment:
This is related to -
* #2645

We should consider the behavior of `--uninstall-previous --all` and probably add a reference to #2645 in the notes of this spec

Reply:
I linked to the issue. Draft away :)